### DISCUSSION TITLE
Thoughts on a better abstract-* API

### DISCUSSION THREAD

**Discussion by sritchie** [sicmutils/sicmutils#399] [Ideas]
I have been having design twitches the past few weeks. These arrived while thinking about how to efficiently implement `ABSTRACT-UP`, `ABSTRACT-DOWN` and friends.

(This is related to the new `symb:and`, `symb:or` and `symb:not`, because those new functions operate on ABSTRACT-BOOLEAN instances, which don't currently have any treatment in the simplifier. So how to build an abstract-bool simplifier, abstract-up simplifier etc and share rules?)

Note that there is a strong connection between

1. Functions that operate on abstract symbolic types (abstract-{number,up,down,matrix,function}, but should also include abstract-{operator,quaternion,real,integer,boolean} etc
2. the construction of simplifiers for each type
3. function compilation, ie, specialization of a symbolic expression down to functions that can operate on concrete instances.
4. TESTS that the concrete implementations of these fns are correct!

For the case of abstract complex numbers:

(1) lives in numsymb.scm
(2) is spread across numsymb.scm, rules.scm and other places where toggles for the simplifier live
(3) only works for expressions of abstract complex numbers; it's not currently possible to compile a function of abstract-matrix operations down to some concrete numerical matrix implementation, for example
(4) doesn't exist yet

I think we can tease apart some ideas combined in numsymb.scm and rules.scm, and make it easy for users to write simplifiers for all sorts of abstract types of different generalities.

Many numsymb.scm functions are combining (non-pattern-matching) simplification rules with expression building. For example:

```scm
(define (symb:conjugate z)
  (cond ((complex? z) (conjugate z))
        ((known-real? z) z)
        ((and (pair? z)
              (memq (operator z)
                    *conjugate-transparent-operators*))
         (cons (operator z)
               (map symb:conjugate (operands z))))
        (else
         `(conjugate ,z))))
```

Is actually a combination of these two ideas:

```scheme
(define (symb:conjugate z)
  (conjugate-rules
   `(conjugate ,z)))

(define (known-transparent? op)
  (memq op *conjugate-transparent-operators*))

(define conjugate-rules
  (rule-system
   ((conjugate (? z complex?))
    none
    (: (conjugate z)))

   ((conjugate (? z known-real?))
    none
    (: z))

   ((conjugate ((? op known-transparent?) (?? xs)))
    none
    ((: op) (conjugate (:: (map conjugate-rules xs)))))))
```

The problem here is that the baked-in simplification rules need to be re-implemented in `rules.scm', which you might not do because of the required duplication. (and the rules in CONJUGATE-RULES are in fact missing in rules.scm.)

Without the rules, you could replace much of numsymb.scm with a form like

```scm
(define-abstract-type
  'number
  ;; Everything passed to addto-symbolic-operator-table... for this to work
  ;; these would need arities, or type signatures if the input is some different
  ;; type
  '(make-rectangular
    make-polar
    real-part
    imag-part
    + - negate
    ;; etc etc
    ))
```

And then some association of a subset of all simplification rules to each constructor.

Add to this the observation that many of the rules we care about can be built out of higher-level combinators, like ASSOCIATIVE, COMMUTATIVE, IDEMPOTENT, UNARY-ELIMINATION, CONSTANT-ELIMINATION etc, and you see how it would become easy to define new abstract types. Here is my simplifier for operators, for example:

```scm
(rule-simplifier
 (associative '+ '*)
 exponent-contract
 (unary-elimination '+ '*))
```

More on this idea below.

An abstract type (I proclaim!) is defined by

- the set of operations that the system should allow between symbolic representations of elements of the type (AND, NOT, OR, IMPLIES, EQUALS etc, for an ABSTRACT-BOOL)

- a set of simplification rules that can operate on symbolic instances of the
type, for example: (NOT (NOT (? x))) => (? x), etc etc

- metadata on each rule, saying WHEN it should run. (options are *at-construction* plus many user-defined toggles, like all of the toggles in rules.scm)

Nothing too special yet. But take these observations:

- If you have a concrete implementation of your type, you get efficient function compilation for free, since compilation ==
- pass symbolic arguments instead of concrete,
- run simplifier on resulting expression
- call eval on (lambda (a,b,c) <simplified-expression>)

- If you can tell the system how to generate concrete instances of your type, every simplification rule comes with an automatic generative test. (Check that expressions before and after each rule application evaluate to the same thing, given many concrete inputs.)

- because many types share operations and simplification rules, we could offer quite a nice API for pre-specifying bundles of rules for the user. Now we are getting into abstract algebra as API.

it is possible to declare very general abstract types like ABSTRACT-SEMIGROUP, ABSTRACT-FIELD etc and have simplifiers for these make sense.

ABSTRACT-SEMIGROUP expressions can only do associative addition:

```scm
(+ (abstract-semigroup 'x)
   (+ (abstract-semigroup 'x)
      (abstract-semigroup 'y)))
#|
(+ x x y)
|#
```

But I can add more structure by "specializing" this expression into ABSTRACT-RING, ABSTRACT-FIELD, or ABSTRACT-UP, ABSTRACT-MATRIX, which will add more allowed operations AND turn on more simplification rules.

**Comment by sritchie** [comment]
@MagusMachinae @light-matters here are some thoughts on the API that would make your project in #270 much easier! Let me know what you think.

**Reply by light-matters** [reply]
Great, I'll have a look soon!

**Reply by light-matters** [reply]
Can't say that I follow everything yet (are sym:and, sym:or etc. already implemented?) but I think using the existing categories of mathematics would make a lot of sense. As well as being appealing in a purity sense, I think it's inevitable that this is where you would end up in the end anyways if you keep trying to abstract everything. So going there consciously in advance allows you to plan for any potential performance hits and really frees the developer to use their existing knowledge (about maths) rather than try and learn your new language and rules.The points about higher level combinators similarly ring true and would be very elegant as well as practical. If I can create new abstract types by pinning terms like associative, commutative etc. to them then the Dirac algebra could be written overnight.

**Comment by sritchie** [comment]
Another note here from GJS...

```
;;; In order to implement derivatives with respect to abstract
;;; quantities we need to create more types -- differential vector,
;;; differential matrix, etc?  Let's attack that later.
```

**Reply by light-matters** [reply]
I think an 'operator' type would be key. This usually takes the form of a matrix of course but it would be better not to tie it down too quickly.
